### PR TITLE
Lazy loading

### PR DESCRIPTION
**Pull request by Alwan**
- Please make sure the below checklist is followed for Pull Requests.

- [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
- [ ] Tests are added where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

Greetings, 

This is my first pull request ever. lazy loading support for admin and entity modules which will improve performance and the start up time.

need help fixing two issues I found:
1- the translation not showing in the lazy loaded modules.
2- the popup dialog not working properly for example url create user dialog 
'/admin/(user-management//popup:user-management-new)'
 
any insight and suggestion ?

### CODE REVIEW

**Comment by deepu105** [comment]
The translation loading changed in the latest master can you try again
after updating your branch with master?

File: generators/client/templates/angular/_package.json [hunk 1]
```diff
@@ -61,6 +61,7 @@
     "@types/selenium-webdriver": "2.53.39",
     <%_ } _%>
     "add-asset-html-webpack-plugin": "1.0.2",
+    "angular-router-loader": "^0.6.0",
```
Comment (deepu105) [inline generators/client/templates/angular/_package.json]: Use fixed version

File: generators/client/templates/angular/src/main/webapp/app/_app.module.ts [hunk 2]
```diff
@@ -46,16 +46,21 @@ import {
     ErrorComponent
 } from './layouts';
 
+const LAZY_ROUTES = [
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/_app.module.ts]: This const can be in lazy.route.ts as exported from there

File: generators/client/templates/angular/src/main/webapp/app/_app.module.ts [hunk 3]
```diff
@@ -46,16 +46,21 @@ import {
...
         BrowserModule,
         LayoutRoutingModule,
         Ng2Webstorage.forRoot({ prefix: 'jhi', separator: '-'}),
+        RouterModule.forRoot(LAZY_ROUTES, { useHash: true }),
         <%=angular2AppName%>SharedModule,
         <%=angular2AppName%>HomeModule,
-        <%=angular2AppName%>AdminModule,
         <%=angular2AppName%>AccountModule,
-        <%=angular2AppName%>EntityModule
+
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/_app.module.ts]: Remove blank line

**Comment by Alwan** [comment]
The translation still not working.

**Comment by Alwan** [comment]
Can you upgrade to 'ngx-translate'? It has better support for lazy loaded modules. You can also set completely different configuration using 'TranslateModule.forChild'.

**Comment by deepu105** [comment]
Ngx translate has some breaking changes which was preventing the upgrade
earlier but since we changed how i18n is loaded I guess we can upgrade and
get rid of the translatePartial loader I implemented to handle partials.
But given my current schedule it might take some time. But if you would
like to do it you are welcome.

**Comment by sendilkumarn** [comment]
yeah agree that we should migrate to ngx-translate... 

@Alwan interested to do a PR ?

**Comment by Alwan** [comment]
I will try. should I open the PR here in jhipster-generator or in ng-jhipster library?

**Comment by deepu105** [comment]
It will be in ng-jhipster

**Comment by sendilkumarn** [comment]
@Alwan working on ngx-translate

**Comment by Alwan** [comment]
I tried to upgrade so many issues I couldn't solve.
I figured the best solution is translation file per module not per component then I faced this bug in ngx-translate 
https://github.com/ngx-translate/core/issues/444

File: generators/client/templates/angular/src/main/webapp/app/_app.module.ts [hunk 4]
```diff
@@ -46,16 +46,16 @@ import {
     ErrorComponent
 } from './layouts';
 
+
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/_app.module.ts]: unnecessary space. remove

File: generators/client/templates/angular/src/main/webapp/app/_lazy.route.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<%#
+ Copyright 2013-2017 the original author or authors.
```
Comment (deepu105) [inline generators/client/templates/angular/src/main/webapp/app/_lazy.route.ts]: Update to latest licence. look at a recent file

**Comment by jdubois** [comment]
I tried to upgrade to ngx-translate, but got so many issues I couldn't fix.... @sendilkumarn @wmarques did you have time to check this out? Otherwise we should stay on the current version, let's not do too many changes at once.

**Comment by deepu105** [comment]
Upgrading to ngx translate will require a complete rewrite of the
translation part on our side. I started it but will take more time. So i
suggest we park these for later as these are not that important at the
moment

**Comment by deepu105** [comment]
ng-jhipster is updated to use latest ngx-translate so can you try with latest master @Alwan ?

**Comment by Alwan** [comment]
I was trying today. 
lazy loading with auxiliary router outlet is not working so popup will not working
related issue
https://github.com/angular/angular/issues/10981

**Comment by sendilkumarn** [comment]
can we use that workaround mentioned there ?

**Comment by Alwan** [comment]
I fixed the popup. I will do the translation later. 
I think we should put a separated translation file for admin module and load it when we open the admin module and likewise for the entity module.
What do you think?

**Comment by deepu105** [comment]
@Alwan I think for the moment lets use the merged single json

**Comment by sendilkumarn** [comment]
@Alwan agree with deepu, we can first target on merged json file.

**Comment by deepu105** [comment]
@Alwan sorry for the delay on this. We are planning to release Angular non-Beta this week and it means we can now target this. Can you please rebase this with the latest master? then I'll review it and we can try to merge it after this week

**Comment by deepu105** [comment]
@Alwan it might be a good time to resume this

**Comment by deepu105** [comment]
I resolved the conflicts. you need to add `angular-router-loader` to webpack[dev|prod] files accordingly as I removed it from common config

**Comment by deepu105** [comment]
@Alwan sorry to keep this on hold, If you are still interested this would be a nice addition indeed. let me know if you need help

**Comment by deepu105** [comment]
@sendilkumarn @wmarques  since there is no response from @Alwan can any one of you continue this?

**Comment by deepu105** [comment]
Seems like this is not going anywhere @sendilkumarn @flaviencathala @wmarques can any of you take this over, I would really hate to waste this

**Comment by sendilkumarn** [comment]
Let me have a look at this. I will first need to see what are the changes that happened in ng5

**Comment by deepu105** [comment]
@NilsWild would you like to give this a shot?

**Comment by NilsWild** [comment]
i'll take a look

**Comment by NilsWild** [comment]
Ok this is complicated with the current structure and the HttpInterceptors as each lazy loaded module currently gets its own instance of HttpClient and thus wont be intercepted. This will take some time

**Comment by sendilkumarn** [comment]
@deepu105 I would like to see the restructuring of folders to happen first.

**Comment by NilsWild** [comment]
@sendilkumarn  i agree. maybe keep in mind that we should keep all shared modules in an core module with forRoot so all lazy modules cann inject those

**Comment by deepu105** [comment]
IMO restructuring wont affect this that much, restructuring is more to
ensure microservice entities are properly grouped and may be to remove use
of baseEntity, nothing drastic


Thanks & Regards,
Deepu

**Comment by deepu105** [comment]
@sendilkumarn @Alwan @NilsWild is anyone interested in moving this forward? if so please comment here else i'm closing this

**Pull request closed without merging**